bouncer/src/repo/manipulator/modelutility/repo_mesh_map_reorganiser.cpp

```diff
@@ -157,7 +157,14 @@ repo::core::model::MeshNode MeshMapReorganiser::getRemappedMesh() const
 			bboxArr.push_back({ bbox[1].x, bbox[1].y, bbox[1].z });
 		}
 
-		auto newMesh = repo::core::model::RepoBSONFactory::makeMeshNode(newVertices, newFaces, newNormals, bboxArr, newUVs, newColors);
+		auto submeshIds = getIDMapArrays();
+		std::vector<float> newIds;
+		for (auto& buf : submeshIds)
+		{
+			newIds.insert(newIds.end(), buf.begin(), buf.end());
+		}
+
+		auto newMesh = repo::core::model::RepoBSONFactory::makeMeshNode(newVertices, newFaces, newNormals, bboxArr, newUVs, newColors, newIds);
```

Comment:
do we really need this id array?

Is this not the same as the ids within the meshMapping?

Reply:
@sebjf 

I guess my issue here is that we're blurring the line between what we declare as a mesh data within the BSON document and what we're using as an interm structure for the mesh batching 🤔 

MeshNode within RepoBsonFactory is suppose to be a 1:1 mapping with the schema that is stored in the database. We "augmented" this for the super meshes which we still store as a mesh in stash.3drepo as a scene structure.

The weird thing here is that we've augmented the bson class to add a property within the database schema, that never hits the database - which is a bit weird. Also the naming of `ids` is confusing as a mesh has `id`, but it's not immediately obvious what `ids` would mean in the context.

I guess what this should really be is a in memory structure that we store as a temporary thing as we construct the repo bundle mappings (that is not a clash of RepoBSONFactory)? Going forward, if we stop storing the stash graph then there isn't really much reason to have the meshMapping within the repo MeshNode even. It would probably be more memory/performance efficient as well, since by not piggy backing onto the mongo library it doesn't have silly enforcements such as immutable properties and storing everything into a serialised buffer etc....

I'm not sure what's the best way forward for this issue? whether we leave it as is and make it a note to refactor when we get rid of the stash graph (In which case I think we should rename ids to `subMeshIds` or something), or we just extrapolate the ids from the mesh mapping, or refactor this all together..

Whatever you feel is best 😆

Reply:
> MeshNode within RepoBsonFactory is suppose to be a 1:1 mapping with the schema that is stored in the database. 

Ah good point, I wasn't really looking at it from this perspective...

I am working on on splitting MeshNode/SupermeshNode as per comments above, so hopefully the discrepancy can be resolved in that.

The question is though is still how to do it. Semantically, the correct thing is to make SupermeshNode subclass MeshNode but then again we have a RepoNode type that never gets written to the database...

Another possibility is to define a new Supermesh type that contains everything MeshNode does,but is in-memory only, and have `getRemappedMesh()` return that?

Then RepoBundleExport can operate on this new in-memory type from then on.

MeshMapReorganiser would also be able to return a MeshNode for the purposes of AssetBundleCreator until it is decommissioned, after which supermeshes would only be handled in memory, and the mappings could be removed from MeshNode entirely?

Reply:
@sebjf 

> The question is though is still how to do it. Semantically, the correct thing is to make SupermeshNode subclass MeshNode but then again we have a RepoNode type that never gets written to the database...

yea it's a bit of an overkill if we are not going to be storing any form of supermesh node soon enough..

> Another possibility is to define a new Supermesh type that contains everything MeshNode does,but is in-memory only, and have getRemappedMesh() return that?

possibly, that's kind of what we do during file imports when we're not quite ready to to construct the MeshNodes, so the behaviour would be similar to other parts of the code.

I think the in memory solution probably works best, but obviously the most rewrite. Not sure whether you feel like it would make more sense to do it when we scrap the stash graph or it makes sense to do it now. I leave this up to you 😄

Reply:
I think its best to split MeshNode, address #678 and decommission AssetBundleCreator & the stash graph in one ticket after RepoBundles have been tested for a while, as they are all interdependent.

My experience over the past few weeks is that trying to make changes while keeping AssetBundleCreator & associated APIs working has usually opened a can of worms, and I expect this to be the same!
If we do it once we don't have to worry about backwards compatibility it will go quicker, and with the increased flexibility will probably end up neater overall.